src/Semi.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE CPP #-}
+
+#if __GLASGOW_HASKELL__ >= 702
+{-# LANGUAGE Safe #-}
+#endif
+
+#if __GLASGOW_HASKELL__ == 708
+{-# OPTIONS_GHC -fno-warn-amp #-}
+#endif
+
+{-|
+
+This module re-exports operators from 'Apply' and 'Bind', but under the same
+names their 'Applicative' and 'Monad' counterparts. This makes it convenient
+to use do-notation on a type that is a 'Bind' but not a monad (or an 'Apply'
+but not an 'Applciative' with @ApplicativeDo@), either using the
+@QualifiedDo@ extension or the more traditional @RebindableSyntax@.
+
+@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE QualifiedDo #-}
+
+foo :: Apply f => f a -> f b -> f (a, b)
+foo as bs = Semi.do
+  a <- as
+  b <- bs
+  pure (a, b)
+
+
+bar :: Bind m => (a -> b -> m c) -> m a -> m b -> m c
+bar f as bs = Semi.do
+  a <- as
+  b <- bs
+  f a b
+@
+
+-}
+
+module Semi
```

Comment:
At the risk of bikeshedding, I do think we should think of a more descriptive name than just `Semi`, especially since it doesn't mesh with the module namespacing that the rest of the package uses. Perhaps we can take inspiration from other modules used for the purposes of assisting `QualifiedDo`, such as `template-haskell`'s [`Language.Haskell.TH.CodeDo`](https://gitlab.haskell.org/ghc/ghc/-/blob/b5155a6c74e774d99a67f2b5c8c06b70ad5e5b2e/libraries/template-haskell/Language/Haskell/TH/CodeDo.hs) module. Perhaps something `Data.Semigroupoid.Do` or `Data.SemiDo`?